### PR TITLE
chore(package): Add ESLint

### PR DESCRIPTION
**Pull request by ExE-Boss**
This adds a&nbsp;basic ESLint configuration to&nbsp;the&nbsp;project which doesn’t require changing almost every file like&nbsp;#1483 (the&nbsp;goal is&nbsp;to&nbsp;slowly enable&nbsp;rules across&nbsp;multiple&nbsp;PRs).

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+		"no-empty": "off",
+		"no-inner-declarations": "off",
+		"no-mixed-spaces-and-tabs": "error",
+		"no-redeclare": "off",
+		"no-sparse-arrays": "off",
+		"no-unused-vars": [
+			"warn",
+			{
+				"args": "none",
+				"argsIgnorePattern": "^_",
```
Comment (RunDevelopment) [inline .eslintrc]: We don't check arguments anyway, so we don't need this pattern.

File: .eslintrc [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+				"varsIgnorePattern": "^_",
```
Comment (RunDevelopment) [inline .eslintrc]: There are no unused variables which match this pattern.

File: gulpfile.js [hunk 3]
```diff
@@ -61,6 +62,13 @@ function minifyJS() {
 	];
 }
 
+function lint(cb) {
+	return pump([
+		src(['**/*.js', '!node_modules/**'], { base: './' }),
+		eslint(),
+		eslint.format(),
+	], cb);
```
Comment (RunDevelopment) [inline gulpfile.js]: Do you really want to not [fail on lint errors](https://github.com/PrismJS/prism/blob/662bbfe5828a0160d2035940469f8a9339fc3f22/gulpfile.js#L53)?

Comment (ExE-Boss) [reply gulpfile.js]: I&nbsp;accidentally forgot to&nbsp;add&nbsp;it.

File: .eslintrc (hunk 2 above)
Comment (ExE-Boss) [reply .eslintrc]: There&nbsp;is&nbsp;now:

https://github.com/PrismJS/prism/blob/3fb78077e23ddc2cdd102506d1a43735764e329e/code.js#L21

Comment (RunDevelopment) [reply .eslintrc]: Hmmm, unused variables are usually a sign that something went wrong during refactoring or writing, so why would we want to ignore these?

(`_`)`after` is a good example: Why is it there? Does it do anything?
`nextSibling` should be side-effect free, so why _not_ remove it. I don't see the benefit in ignoring a potential mistake because of its name.
(I mean, even after the ESLint warning disappears, I have like 3 other linters which all scream at me to remove it...)

Also, even after you remove the ignore pattern, `after` stays the only unused variable.

File: .eslintrc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+				"Promise": "readonly",
+				"components": "readonly",
+				"setTheme": "readonly",
+				"PrefixFree": "readonly",
+				"StyleFix": "readonly",
+			},
+			"rules": {
+				"no-cond-assign": "off",
+				"no-mixed-spaces-and-tabs": "off",
+				"semi": "off",
```
Comment (RunDevelopment) [inline .eslintrc]: Why did you turn this off?

(There are 2 missing semicolons in these files.)

Comment (ExE-Boss) [reply .eslintrc]: Ok, I’ll&nbsp;fix&nbsp;it.

I&nbsp;originally tried to&nbsp;avoid changing `code.js` and&nbsp;`downloads.js` too&nbsp;much.

File: .eslintrc [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+			"env": {
+				"es6": true,
+			},
+			"parserOptions": {
+				"ecmaVersion": 2015,
+			},
+			"rules": {
+				"comma-dangle": [
+					"error",
+					"always-multiline",
```
Comment (RunDevelopment) [inline .eslintrc]: Why did you change the behavior here?

File: .eslintrc [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+				],
+			},
+		},
+		{
+			"files": [
+				"code.js",
+				"download.js",
+				"examples.js",
+				"utopia.js",
```
Comment (RunDevelopment) [inline .eslintrc]: I'm not really sure if we should lint this file. It feels more like a library to me.

@LeaVerou Isn't this the predecessor to [Bliss](https://github.com/LeaVerou/bliss)?

Comment (ExE-Boss) [reply .eslintrc]: Yes, this is the predecessor to Bliss, but it’s not a 100% drop&#x2011;in replacement.

File: .eslintrc (hunk 5 above)
Comment (ExE-Boss) [reply .eslintrc]: Because `gulpfile.js` already has traling commas in&nbsp;most places.

Comment (RunDevelopment) [reply .eslintrc]: But do we really want to force people to use commas because of that?
Also, isn't ESLint support to make the code style consistent?

File: .eslintrc (hunk 6 above)
Comment (RunDevelopment) [reply .eslintrc]: Wait. Don't we ignore utopia anyway? (`.eslintignore`)

File: .eslintrc (hunk 2 above)
Comment (RunDevelopment) [reply .eslintrc]: I think we can just remove the unused variable.

File: .eslintrc (hunk 6 above)
Comment (ExE-Boss) [reply .eslintrc]: Right, we&nbsp;do…

File: .eslintrc (hunk 5 above)
Comment (ExE-Boss) [reply .eslintrc]: I eventually want to enforce <code>[comma‑dangle](https://eslint.org/docs/rules/comma-dangle): "always‑multiline"</code> everywhere.

Comment (RunDevelopment) [reply .eslintrc]: I can see why you want this but it literally requires us to change (almost) every language definition.

I'm personally quite a fan of `"only-multiline"`.
We should get some more opinions on this.

@mAAdhaTTah @Golmote @LeaVerou

File: .eslintrc (hunk 2 above)
Comment (RunDevelopment) [reply .eslintrc]: Can't we remove the ignore pattern now or is there a reason to keep it?

File: code.js [hunk 7]
```diff
@@ -198,7 +198,7 @@ function listPlugins(ul) {
 		$u.element.create('li', {
 			contents: {
 				tag: 'a',
-				prop: {
+				properties: {
```
Comment (RunDevelopment) [inline code.js]: Nice catch!

File: components/prism-core.js [hunk 8]
```diff
@@ -1,7 +1,7 @@
+/* global WorkerGlobalScope */
```
Comment (RunDevelopment) [inline components/prism-core.js]: I get that it's convenient to just declare it here but wouldn't it be better to make an overwrite in `.eslintrc` for `components/prism-core.js` and `prism.js` to add this global variable?

File: examples.js [hunk 9]
```diff
@@ -166,8 +166,8 @@ function update(id) {
  * Loads a language, including all dependencies
  *
  * @param {string} lang the language to load
- * @type {Promise} the promise which resolves as soon as everything is loaded
- */
+ * @returns {Promise<Event|void>} the promise which resolves as soon as everything is loaded
+s */
```
Comment (RunDevelopment) [inline examples.js]: s?

File: examples.js [hunk 10]
```diff
@@ -166,8 +166,8 @@ function update(id) {
  * Loads a language, including all dependencies
  *
  * @param {string} lang the language to load
- * @type {Promise} the promise which resolves as soon as everything is loaded
- */
+ * @returns {Promise<Event|void>} the promise which resolves as soon as everything is loaded
```
Comment (RunDevelopment) [inline examples.js]: Missing `.` after `Promise`?

File: package.json [hunk 11]
```diff
@@ -23,8 +23,10 @@
   },
   "devDependencies": {
     "chai": "^4.2.0",
+    "eslint": "^5.15.1",
```
Comment (RunDevelopment) [inline package.json]: Isn't just `gulp-eslint` enough? Why do we also need `eslint` itself?

Comment (ExE-Boss) [reply package.json]: Because `gulp‑eslint` declares `eslint` as&nbsp;a&nbsp;`peerDependency` (or&nbsp;at&nbsp;least it&nbsp;did when I&nbsp;last checked)

File: examples.js (hunk 9 above)
Comment (ExE-Boss) [reply examples.js]: Typo

File: examples.js (hunk 10 above)
Comment (ExE-Boss) [reply examples.js]: It’s unnecessary, also I use the [TypeScript style syntax for JSDoc types](https://github.com/Microsoft/TypeScript/wiki/JsDoc-support-in-JavaScript).

File: package.json (hunk 11 above)
Comment (RunDevelopment) [reply package.json]: [ES-lint is a normal dependency](https://github.com/adametry/gulp-eslint/blob/master/package.json#L43) now.

File: examples.js (hunk 10 above)
Comment (RunDevelopment) [reply examples.js]: Seems like JSDoc is also able to handle this syntax.

File: gulpfile.js [hunk 12]
```diff
@@ -184,14 +193,14 @@ function changelog(cb) {
...
 }
 
 const components = minifyComponents;
 const plugins = series(languagePlugins, minifyPlugins);
 
-
+exports.lint = lint;
 exports.watch = watchComponentsAndPlugins;
-exports.default = parallel(components, plugins, componentsJsonToJs, build);
+exports.default = series(parallel(components, plugins, componentsJsonToJs, build), lint);
```
Comment (RunDevelopment) [inline gulpfile.js]: Why can't we do it in parallel?

Comment (ExE-Boss) [inline gulpfile.js]: I’ve made the&nbsp;`lint` task depend on&nbsp;everything in&nbsp;`default`, since if&nbsp;the&nbsp;`prism.js` file hasn’t finished building yet and the&nbsp;`lint` task tries to&nbsp;lint&nbsp;it while it’s&nbsp;being&nbsp;built, it will cause the&nbsp;build to&nbsp;fail.

File: code.js (hunk 7 above)
Comment (ExE-Boss) [reply code.js]: The `prop` alias is&nbsp;only available as&nbsp;a&nbsp;fallback when passed to&nbsp;the&nbsp;`$u.element.set(…)`&nbsp;method (it’s&nbsp;ignored&nbsp;in&nbsp;`$u.element.create(…)`, because `create(…)` always creates a&nbsp;`properties`&nbsp;key).

**Comment by LeaVerou** [comment]
As I have expressed in the other issue, I disagree with any linting being classed as "error". It should be warnings, and errors should be reserved for actual JS errors. A wrong indent is not the same importance as a JS parse error that breaks the whole script.

File: .eslintrc (hunk 5 above)
Comment (LeaVerou) [reply .eslintrc]: I'm not a fan of dangling commas. I see the benefit in multiline cases, but they are also less natural and I think optimizing our code for tooling is a slippery slope. But my objection is not so strong that is should hold this up if the rest of you agree. If you don't however, do count it as an objection.

**Comment by mAAdhaTTah** [comment]
@LeaVerou Is that to suggest we can merge in PRs with warnings? We could switch warnings -> errors in CI if not, so they're just warnings during development.

**Comment by LeaVerou** [comment]
Oh I wasn't even thinking about CI. I primarily find ESLint useful as an authoring environment plugin that highlights errors and warnings inline. I would find it frustrating if it highlighted syntax errors and stylistic issues the same way.

File: gulpfile.js (hunk 12 above)
Comment (mAAdhaTTah) [reply gulpfile.js]: We probably shouldn't lint `prism.js`, as that's a generated file. That said we should probably not build if `lint` errors, right?

**Comment by mAAdhaTTah** [comment]
Sounds good, so we should bump warnings -> errors in CI and downgrade style issues to warnings during development.

**Comment by ExE-Boss** [comment]
And how do you propose to do that?

**Comment by mAAdhaTTah** [comment]
[This might work](https://github.com/eslint/eslint/issues/2309#issuecomment-219828044). Another suggestion is to use a `.js` file for config and use env vars. There's a package called `is-ci` that could help here.

**Comment by RunDevelopment** [comment]
We could also check the output of [eslint](https://www.npmjs.com/package/gulp-eslint#eslintresultsaction) and to if we got any warnings.

File: components/prism-core.js [hunk 13]
```diff
@@ -549,5 +548,5 @@ if (typeof module !== 'undefined' && module.exports) {
 
 // hack for components to work correctly in node.js
 if (typeof global !== 'undefined') {
-	global.Prism = Prism;
+	global['Prism'] = Prism;
```
Comment (mAAdhaTTah) [inline components/prism-core.js]: Not a fan of whatever's causing this. I saw this below as well.

Comment (RunDevelopment) [reply components/prism-core.js]: That's a dirty hack to remove an error message from [TSCheck](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html).

Not a fan either btw.

File: .eslintrc (hunk 5 above)
Comment (mAAdhaTTah) [reply .eslintrc]: I kinda like dangling commas, but in this case, I think the churn on the language definitions makes it not worth it.

I'd also prefer not to have a different rule for the `gulpfile.js` but instead make the changes in the `gulpfile.js` to bring it in line with the rest of the codebase.

File: .eslintrc [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+					"always-multiline",
+				],
+			},
+		},
+		{
+			"files": [
+				"code.js",
+				"download.js",
+				"examples.js",
+			],
```
Comment (mAAdhaTTah) [inline .eslintrc]: I believe these files will need to change as a result of #1781.

Comment (ExE-Boss) [reply .eslintrc]: Yeah, they will…

File: .eslintrc (hunk 2 above)
Comment (mAAdhaTTah) [reply .eslintrc]: Agreed, remove the ignore pattern.

File: .eslintrc [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+			"only-multiline",
+		],
+//		"indent": [
+//			"error",
+//			"tab",
+//			{
+//				"SwitchCase": 1,
+//				"outerIIFEBody": 0,
+//			}
+//		],
```
Comment (mAAdhaTTah) [inline .eslintrc]: There are a couple commented out rules that should be cleaned up.

File: .eslintignore [hunk 16]
```diff
@@ -0,0 +1,4 @@
+*.min.js
+/components.js
+/utopia.js
+/vendor/**
```
Comment (mAAdhaTTah) [inline .eslintignore]: This file will also need to change because of #1781.

File: tests/helper/test-case.js [hunk 17]
```diff
@@ -30,9 +30,6 @@ var TokenStreamTransformer = require("./token-stream-transformer");
  * If the file contains more than three parts, the remaining parts are just ignored.
  * If the file however does not contain at least two parts (so no expected token stream),
  * the test case will later be marked as failed.
- *
- *
- * @type {{runTestCase: Function, transformCompiledTokenStream: Function, parseTestCaseFile: Function}}
```
Comment (mAAdhaTTah) [inline tests/helper/test-case.js]: I don't think this should be removed?

File: download.js [hunk 18]
```diff
@@ -558,8 +560,8 @@ function generateCode(){
 }
 
 function buildCode(promises) {
-	var i = 0,
-	    l = promises.length;
+	var i = 0;
+	var l = promises.length;
```
Comment (mAAdhaTTah) [inline download.js]: These files all have changes which add `var` statements, despite the prism-core files all allowing multiline vars. What's the story here?

Comment (ExE-Boss) [reply download.js]: This is because of [`no-mixed-spaces-and-tabs`](https://eslint.org/docs/rules/no-mixed-spaces-and-tabs) (see commit https://github.com/PrismJS/prism/pull/1796/commits/53a54b40a433ca390918a7d3d0514718fc438ac8).

I’ll probably use the `smart-tabs` option which I didn’t know about until now.

File: tests/helper/test-case.js (hunk 17 above)
Comment (RunDevelopment) [reply tests/helper/test-case.js]: It should because the typing is incomplete and only gets in the way of TS's type inference and the doc/types provided by the JSDoc of the functions.

File: download.js (hunk 18 above)
Comment (RunDevelopment) [reply download.js]: [Here](https://github.com/PrismJS/prism/pull/1796#pullrequestreview-212046340)'s the story.

The problem is that tabs followed by spaces are considered mixed indentation and some auto formatter just replace the 4 spaces with a tab. Most of Prism core has already fallen victim to these formatters. But instead of saving the wounded I think it's easier to just finish them off. (Enough metaphors for today)

The idea is that if you just write the additional `var`s you don't have these problems (mixed indentation, war against formatters).

File: .eslintrc (hunk 5 above)
Comment (ExE-Boss) [reply .eslintrc]: The thing is that some code is written as if [`comma‑dangle`](https://eslint.org/docs/rules/comma-dangle) was `"always‑multiline"` and other as if it was `"only‑multiline"`.

**Comment by LeaVerou** [comment]
I saw some back and forth on dangling commas. My opinion is that if there is not near 100% consensus on a linting rule, I'd suggest not enforcing one either way. We don't need to control every single aspect of code style via linting, especially if they haven't been enforced so far at all. We don't need to provide a value for every single rule ESLint supports. Linting should help our work, not get in the way, and overly strict linting can easily get in the way.
Just my 2 cents, take it or leave it!

File: gulpfile.js [hunk 19]
```diff
@@ -59,10 +60,18 @@ function inlineRegexSource() {
...
+		uglify(),
 	];
 }
 
+function lint(cb) {
+	return pump([
+		src(['**/*.js', '!node_modules/**'], { base: './' }),
+		eslint(),
+		eslint.format(),
+		eslint.failAfterError(),
```
Comment (RunDevelopment) [inline gulpfile.js]: Implementing a `failAfterWarningOrError` should be trivial looking at the [implementation of `failAfterError`](https://github.com/adametry/gulp-eslint/blob/master/index.js#L161).

**Comment by RunDevelopment** [comment]
@ExE-Boss Sorry, I did expect to be able to push to your branch. I reverted everything and make a PR.

**Comment by RunDevelopment** [comment]
The only things left to discuss now are:

1. How to handle [joint variable declarations](https://github.com/PrismJS/prism/pull/1796#pullrequestreview-212046340).
    Do we want to leave the tabs as is or should we add a few more `var`s.
    Using spaces is not possible because of `"no-mixed-spaces-and-tabs": "error"`. [Smart tabs](https://eslint.org/docs/rules/no-mixed-spaces-and-tabs#smart-tabs) are possible but a real pain because some formatters (e.g. the one I'm using) don't support them.
2. The [TSC hack](https://github.com/PrismJS/prism/pull/1796#discussion_r268459515).

File: .eslintrc [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+		],
+		"linebreak-style": [
+			"error",
+			"unix",
+		],
+		"no-console": "off",
+		"no-empty": "off",
+		"no-inner-declarations": "off",
+		"no-mixed-spaces-and-tabs": "error",
+		"no-redeclare": "off",
```
Comment (RunDevelopment) [inline .eslintrc]: I think that this should be set to `"error"`.

When we switch to ES6 and replace all `var`s with either `const` or `let` in the future, we will have to deal with these errors anyway, so we might as well do it now.

File: .eslintrc [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+		"comma-dangle": [
+			"warn",
+			"only-multiline",
+		],
+		"linebreak-style": [
+			"error",
+			"unix",
+		],
+		"no-console": "off",
+		"no-empty": "off",
```
Comment (RunDevelopment) [inline .eslintrc]: An empty block is usually a sign that you forgot to do something, so I think that this should be a warning.

One notable exception is [empty `catch` blocks](https://eslint.org/docs/rules/no-empty#allowemptycatch).

```suggestion
		"no-empty": [
			"warn",
			{
				"allowEmptyCatch": true
			}
		],
```

File: plugins/jsonp-highlight/prism-jsonp-highlight.js [hunk 22]
```diff
@@ -39,15 +39,16 @@
 	 * @returns {Adapter} A registered adapter or `null`.
 	 */
 	function getAdapter(adapter) {
+		var i = 0, item;
```
Comment (RunDevelopment) [inline plugins/jsonp-highlight/prism-jsonp-highlight.js]: You don't need to initialize `i`. The for loops do that already.

```suggestion
		var i, item;
```

File: plugins/command-line/prism-command-line.js [hunk 23]
```diff
@@ -34,9 +34,11 @@ Prism.hooks.add('before-highlight', function (env) {
 
 	var outputSections = pre.getAttribute('data-output');
 	var outputFilter = pre.getAttribute('data-filter-output');
+
+	var i = 0;
```
Comment (RunDevelopment) [inline plugins/command-line/prism-command-line.js]:
```suggestion
	var i;
```

**Comment by mAAdhaTTah** [comment]
Ooof, letting this lag will make it very difficult to merge. Are we all aligned on the rules presented here? I might go so far as to suggest slotting this into a potential 2.0 and introduce it in one sweeping change early on there. Thoughts?

**Comment by LeaVerou** [comment]
>  Using spaces is not possible because of `"no-mixed-spaces-and-tabs": "error"`. [Smart tabs](https://eslint.org/docs/rules/no-mixed-spaces-and-tabs#smart-tabs) are possible but a real pain because some formatters (e.g. the one I'm using) don't support them.

We should never use tabs for alignment. The codebase uses smart tabs. If we don't want to enforce them with ESLint we should avoid any rules related to indentation for now, not inserting rules that will force us to worsen how we do indentation and alignment! Again, we don't need to lint everything. We can start with a small set of rules and expand as we go.

(Also, if a tool doesn't support smart tabs, I'd consider that a bug and open an issue with them)

**Comment by RunDevelopment** [comment]
About alignment: Why don't we use Prettier and end the formatting debate?

**Comment by LeaVerou** [comment]
Does Prettier support smart tabs?

**Comment by RunDevelopment** [comment]
Just checked. It doesn't and I'd even argue that it doesn't matter. I have yet to come across a formatter that handles smart tabs correctly.
Examples: ESLint doesn't (see above). Clang format (C/C++ formatter that can also do JS) can kinda do it but it assumes that tabs are 8 spaces wide, so the alignment is all messed up with my usual 4 spaces = 1 tab. VSCode's builtin formatter can't do it.

Point is: Which formatter has good support for smart tabs? If there aren't any good formatters that support it, then I don't think this should be a criterion. Manual formatting just isn't feasible and no formatter is a lot worse than a formatter that uses, god forbid, tabs for alignment, IMO.

Don't get me wrong: I think you're correct. Tabs for indentation and spaces for alignment is the indented use for tabs and spaces but if all our automated tools work against us on this matter, it might be easiest to just not care about it.

**Comment by RunDevelopment** [comment]
Closed in favor of #2831.
